### PR TITLE
Feature/bimpm

### PR DESCRIPTION
**Pull request by rgtjf**
https://github.com/NTMC-Community/MatchZoo-py/issues/28

### CODE REVIEW

File: matchzoo/models/bimpm.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
+"""An implementation of ESIM Model."""
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: BiMPM

File: matchzoo/models/bimpm.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+        """:return: Default padding callback."""
+        return callbacks.BasicPadding()
+
+    def build(self):
+        """Make function layers."""
+        self.embedding = None
+        self.context_LSTM = None
+        self.aggregation_LSTM = None
+        self.pred_fc1 = None
+        self.pred_fc2 = None
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: I think Line 55-59 is redundant

File: matchzoo/models/bimpm.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            batch_first=True
+        )
+
+        # Prediction Layer
+        self.pred_fc1 = nn.Linear(
+            self._params['hidden_size'] * 4,
+            self._params['hidden_size'] * 2)
+        # self.pred_fc2 = nn.Linear(
+        #     self._params['hidden_size'] * 2,
+        #     self.args.class_size)
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Delete useless comments

File: matchzoo/models/bimpm.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        self.pred_fc2 = self._make_output_layer(
+            self._params['hidden_size'] * 2)
+
+        # parameters
+        self.reset_parameters()
+
+    def forward(self, inputs):
+        """Forward."""
+
+        self.num_psp = self._params['num_perspective']
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Move this to ```build()```?

File: matchzoo/models/bimpm.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+
+        # Matching Layer
+        def mp_matching_func(v1, v2, w):
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Defining functions should be outside of `forward()`

File: matchzoo/models/bimpm.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+class BiMPM(BaseModel):
+    """
+    BiMPM Model.
+
+    Reference:
+    - https://github.com/galsang/BIMPM-pytorch/blob/master/model/BIMPM.py
+
+    Examples:
+        >>> model = BiMPM()
+        >>> model.guess_and_fill_missing_params(verbose=0)
```
Comment (faneshion) [inline matchzoo/models/bimpm.py]: Please add parameter configuration details, as the example would be a tutorial for this model.

File: matchzoo/models/bimpm.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+        """:return: model default parameters."""
+        params = super().get_default_params(
+            with_embedding=True,
+            with_multi_layer_perceptron=False
+        )
+        params.add(Param(name='mask_value', value=0,
+                         desc="The value to be masked from inputs."))
+        params.add(Param(name='dropout', value=0.2,
+                         desc="Dropout rate."))
+        params.add(Param(name='hidden_size', value=100,
```
Comment (faneshion) [inline matchzoo/models/bimpm.py]: For hyper-parameters which can be tuned in the auto ML component, you can add `hyper_space ` in the Param definition. You can refer to 'matchzoo/models/conv_knrm.py' to see how to add them.

File: matchzoo/models/bimpm.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+    Basic mp_matching_func.
+
+    :param v1: (batch, seq_len, hidden_size)
+    :param v2: (batch, seq_len, hidden_size) or (batch, hidden_size)
+    :param w: (num_psp, hidden_size)
+    :return: (batch, num_psp)
+    """
+
+    """
+    # Trick for large memory requirement
```
Comment (faneshion) [inline matchzoo/models/bimpm.py]: Should these lines be removed?

File: matchzoo/models/bimpm.py [hunk 9]
```diff
@@ -0,0 +1,420 @@
+"""An implementation of BiMPM Model."""
```
Comment (faneshion) [inline matchzoo/models/bimpm.py]: Please add this mode in the `__init__.py`, and please run and test it on a specific dataset.

File: matchzoo/models/bimpm.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+        # Word Representation Layer
+        # (batch, seq_len) -> (batch, seq_len, word_dim)
+
+        # [B, L], [B, R]
+        p, h = inputs['text_left'].long(), inputs['text_right'].long()
+
+        # [B, L]
+        # [B, R]
+        # p_mask = (p == self._params['mask_value'])
+        # h_mask = (h == self._params['mask_value'])
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Should Line 100-103 be removed?

File: matchzoo/models/bimpm.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+        # (batch, seq_len, hidden_size->1)
+        v1_norm = v1.norm(p=2, dim=2, keepdim=True)
+        v2_norm = v2.norm(p=2, dim=2, keepdim=True)
+        # (batch, seq_len1, seq_len2)
+        n = torch.matmul(v1, v2.permute(0, 2, 1))
+        d = v1_norm * v2_norm.permute(0, 2, 1)
+        m.append(div_with_small_value(n, d))
+    # list of (batch, seq_len1, seq_len2)
+    #     -> (batch, seq_len1, seq_len2, l)
+    m = torch.stack(m, dim=3)
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Are Lines 350-366 examples or something?

File: matchzoo/models/bimpm.py [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+    return div_with_small_value(a, d)
+
+
+def div_with_small_value(n, d, eps=1e-8):
+    """
+    Small values are replaced by 1e-8 to prevent it from exploding.
+
+    :param n:
+    :param d:
+    :return: n/d
```
Comment (Chriskuei) [inline matchzoo/models/bimpm.py]: Explanations of `n`, `d`, `n/d`

**Pull request merged**